core/src/main/scala/kafka/message/ByteBufferMessageSet.scala

```diff
@@ -226,48 +338,198 @@ class ByteBufferMessageSet(val buffer: ByteBuffer) extends MessageSet with Loggi
+    var sizeInBytesAfterConversion = shallowValidBytes
+    this.internalIterator(isShallow = true).foreach(messageAndOffset =>
+      sizeInBytesAfterConversion += Message.headerSizeDiff(messageAndOffset.message.magic, toMagicValue))
+    val newBuffer = ByteBuffer.allocate(sizeInBytesAfterConversion)
+    var newMessagePosition = 0
+    this.internalIterator(isShallow = true).foreach {messageAndOffset =>
+      val message = messageAndOffset.message
+      validateMessageKey(message, compactedTopic)
+      newBuffer.position(newMessagePosition)
+      // write offset.
+      newBuffer.putLong(offsetCounter.getAndIncrement)
+      // Write new message size
+      val newMessageSize = message.size + Message.headerSizeDiff(message.magic, toMagicValue)
+      newBuffer.putInt(newMessageSize)
+      // Create new message buffer
+      val newMessageBuffer = newBuffer.slice()
+      newMessageBuffer.limit(newMessageSize)
+      // Convert message if necessary
+      if (message.magic != toMagicValue)
+        message.convertToBuffer(toMagicValue, newMessageBuffer)
+      else
+        newMessageBuffer.put(message.buffer.array(), message.buffer.arrayOffset(), message.size)
+
+      newMessagePosition += MessageSet.LogOverhead + newMessageSize
+    }
+    newBuffer.rewind()
+    new ByteBufferMessageSet(newBuffer)
+  }
+
+  private def validateNonCompressedMessagesAndAssignOffsetInPlace(offsetCounter: AtomicLong,
+                                                                  compactedTopic: Boolean,
+                                                                  timestampType: TimestampType,
+                                                                  timestampDiffMaxMs: Long): ByteBufferMessageSet = {
+    // do in-place validation and offset assignment
+    var messagePosition = 0
+    buffer.mark()
+    while (messagePosition < sizeInBytes - MessageSet.LogOverhead) {
+      buffer.position(messagePosition)
+      buffer.putLong(offsetCounter.getAndIncrement())
+      val messageSize = buffer.getInt()
+      val messageBuffer = buffer.slice()
+      messageBuffer.limit(messageSize)
+      val message = new Message(messageBuffer)
+      validateMessageKey(message, compactedTopic)
+      if (message.magic > Message.MagicValue_V0)
+        validateTimestamps(message, timestampType, timestampDiffMaxMs, NoCompressionCodec)
+      // update CRC for the message if the timestamp is LogAppendTime
+      if (timestampType == TimestampType.LogAppendTime)
+        Utils.writeUnsignedInt(messageBuffer, Message.CrcOffset, message.computeChecksum())
+      messagePosition += MessageSet.LogOverhead + messageSize
+    }
+    buffer.reset()
+    this
+  }
+
+  private def validateMessageKey(message: Message, compactedTopic: Boolean) {
+    if (compactedTopic && !message.hasKey)
+      throw new InvalidMessageException("Compacted topic cannot accept message without key.")
+  }
+
+  private def validateTimestamps(message: Message,
```

Comment:
It would be useful to have a comment describing what this method does. Especially because in one "invalid" case it throws exception, and in another case (when we need to overwrite timestamp) it returns true/false.